From 66aa48703ff78c308097871c84ad3946bc2ac6d4 Mon Sep 17 00:00:00 2001
From: "Allen.Ye" <allen.ye@mediatek.com>
Date: Thu, 30 Nov 2023 14:01:29 +0800
Subject: [PATCH 037/137] mtk: mac80211: Fix SMPS action frame cap check

Fix SMPS action frame cap check.
Due to 6G band doesn't have HT cap, we change cap check into each action
frame section.

Signed-off-by: Allen.Ye <allen.ye@mediatek.com>
---
 net/mac80211/rx.c | 5 +++++
 1 file changed, 5 insertions(+)


--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -3666,6 +3666,11 @@
 			struct ieee80211_supported_band *sband;
 			enum ieee80211_smps_mode smps_mode;
 			struct sta_opmode_info sta_opmode = {};
+
+			if (rx->link_sta->pub->he_cap.has_he &&
+			    !(rx->link_sta->pub->he_cap.he_cap_elem.mac_cap_info[5] &
+			    IEEE80211_HE_MAC_CAP5_HE_DYNAMIC_SM_PS))
+				goto invalid;
 
 			if (sdata->vif.type != NL80211_IFTYPE_AP &&
 			    sdata->vif.type != NL80211_IFTYPE_AP_VLAN)
